cmd/mxnet-operator.v1/app/server.go

```diff
@@ -0,0 +1,208 @@
+
+	unstructuredInformer := controller.NewUnstructuredMXJobInformer(kcfg, opt.Namespace)
+
+	// Create mx controller.
+	tc := controller.NewMXController(unstructuredInformer, kubeClientSet, mxJobClientSet, kubeBatchClientSet, kubeInformerFactory, mxJobInformerFactory, *opt)
+
+	// Start informer goroutines.
+	go kubeInformerFactory.Start(stopCh)
+
+	// We do not use the generated informer because of
+	// go mxJobInformerFactory.Start(stopCh)
+	go unstructuredInformer.Informer().Run(stopCh)
+
+	// Set leader election start function.
+	run := func(<-chan struct{}) {
+		if err := tc.Run(opt.Threadiness, stopCh); err != nil {
+			log.Errorf("Failed to run the controller: %v", err)
+		}
+	}
+
+	id, err := os.Hostname()
+	if err != nil {
+		return fmt.Errorf("failed to get hostname: %v", err)
+	}
+
+	// Prepare event clients.
+	eventBroadcaster := record.NewBroadcaster()
+	if err = v1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("CoreV1 Add Scheme failed: %v", err)
+	}
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "mxnet-operator"})
+
+	rl := &resourcelock.EndpointsLock{
+		EndpointsMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      "mxnet-operator",
+		},
+		Client: leaderElectionClientSet.CoreV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity:      id,
+			EventRecorder: recorder,
+		},
+	}
+
+	// Start leader election.
+	election.RunOrDie(election.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: leaseDuration,
+		RenewDeadline: renewDuration,
+		RetryPeriod:   retryPeriod,
+		Callbacks: election.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				log.Fatalf("leader election lost")
+			},
+		},
+	})
+
+	return nil
+}
+
+func createClientSets(config *restclientset.Config) (kubeclientset.Interface, kubeclientset.Interface, mxjobclientset.Interface, kubebatchclient.Interface, error) {
+
+	crdClient, err := crdclient.NewForConfig(config)
+
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	checkCRDExists(crdClient, mxnetv1.MXCRD)
+
+	kubeClientSet, err := kubeclientset.NewForConfig(restclientset.AddUserAgent(config, "mxnet-operator"))
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	leaderElectionClientSet, err := kubeclientset.NewForConfig(restclientset.AddUserAgent(config, "leader-election"))
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	mxJobClientSet, err := mxjobclientset.NewForConfig(config)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	kubeBatchClientSet, err := kubebatchclient.NewForConfig(restclientset.AddUserAgent(config, "kube-batch"))
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	return kubeClientSet, leaderElectionClientSet, mxJobClientSet, kubeBatchClientSet, nil
+}
+
+func checkCRDExists(clientset crdclient.Interface, crdName string) {
+	_, err := clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crdName, metav1.GetOptions{})
```

Comment:
This API call may require cluster level RBAC https://github.com/kubeflow/tf-operator/issues/929

I think we could follow https://github.com/kubeflow/tf-operator/pull/961/files to avoid it.

Reply:
Done